python/hail/docs/tutorials/03-expressions.ipynb

```diff
@@ -0,0 +1,398 @@
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "slideshow": {
+     "slide_type": "slide"
+    }
+   },
+   "source": [
+    "Hail `Expression` objects have an (unsupported) method to print the expression they record."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {
+    "slideshow": {
+     "slide_type": "fragment"
+    }
+   },
+   "outputs": [],
+   "source": [
+    "one = hl.int32(1)\n",
+    "three = one + 2\n",
+    "three._ast.to_hql()"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "slideshow": {
+     "slide_type": "slide"
+    }
+   },
+   "source": [
+    "Hail evaluates an expression only when it must, for example:\n",
+    "\n",
+    " - when performing an aggregation,\n",
+    " - when calling `take`, `collect` or `show`,\n",
+    " - when exporting or writing to disk.\n",
+    "\n",
+    "Hail evaluates expressions by streaming to accomodate very large datasets."
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "You can expressions with no index by calling `value`. The `show` method also prints the type."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "three.value"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {
+    "slideshow": {
+     "slide_type": "fragment"
+    }
+   },
+   "outputs": [],
+   "source": [
+    "three.show()"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "slideshow": {
+     "slide_type": "slide"
+    }
+   },
+   "source": [
+    "# Indices\n",
+    "\n",
+    "Expressions carry another piece of information: indices.  Indices record the `Table` or `MatrixTable` the expression refers to, and what axes the `MatrixTable` varies over.\n",
+    "\n",
+    "Let's see some examples."
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "slideshow": {
+     "slide_type": "slide"
+    }
+   },
+   "source": [
+    "Let's load the 1000 genomes dataset again."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "hl.utils.get_1kg('data/')"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {
+    "slideshow": {
+     "slide_type": "fragment"
+    }
+   },
+   "outputs": [],
+   "source": [
+    "mt = hl.read_matrix_table('data/1kg.mt')\n",
+    "mt"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "slideshow": {
+     "slide_type": "slide"
+    }
+   },
+   "source": [
+    "Let's add a global field."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {
+    "slideshow": {
+     "slide_type": "fragment"
+    }
+   },
+   "outputs": [],
+   "source": [
+    "mt = mt.annotate_globals(dataset = '1kg')"
```

Comment:
that may be Python's standard style for keyword args, but we've been using spaces, which look better in most cases. I think we can view this as using the style rules about assignment rather than keyword args in the annotate methods that are assigning new fields.